### DISCUSSION TITLE
Math Errors in alsa backend?

### DISCUSSION THREAD

**Discussion by JasonLG1979** [librespot-org/librespot#788] [General]
[Here](https://github.com/librespot-org/librespot/blob/dev/playback/src/audio_backend/alsa.rs#L60-L62) the intention seems to be to find the period size:
```rust
    let mut period_size = ((SAMPLES_PER_SECOND * format.size() as u32) as f32
        * (BUFFERED_LATENCY.as_secs_f32() / BUFFERED_PERIODS as f32))
        as Frames;
```

Putting this after it yields `Period Size: 5512` for `S16`:
```rust
println!("Period Size: {}", period_size as usize);
```

But it yields `8268` for `S24_3`, `11025` for `S24` and `11025` for `S32`.

The scale used seems to be 1 byte is assumed to be 2756 for some reason? At least that's what the above equation works out to be.

`S16` = 2 bytes, 2 *  2756 = 5512, `S24_3` = 3 bytes, 3 *  2756 = 8268, `S24` and `S32` = 4 bytes, 4 * 2756 = 11025.

[An ALSA Frame = a sample](https://alsa.opensrc.org/Frame), so 44100 audio has 44100 Frames per sec, it has nothing to do with the format, given the same sampling rate it should not change. If you want a period time of 125ms that would mean  5512.5 (44100/8) (ALSA will round that up to 5513) so the math kinda checks out for `S16` but it should not grow as you up the format.

The end result is that the `period_size` and `buffer_size` (as per `cat /proc/asound/card*/pcm*p/sub*/hw_params`) are not set to correct values on formats other than `S16` and `self.buffer` in `alsa.rs` ends up being ~~up to twice the size as it really needs to be~~ the wrong size.

I have a [wip branch](https://github.com/JasonLG1979/librespot/tree/alsa_backend_fix) I'm working on that I think is a better solution. @roderickvd if you could take a look at it and tell me that you think that would be great?

**Comment by JasonLG1979** [comment]
`cat /proc/asound/card*/pcm*p/sub*/hw_params` Should look like this given a buffer consisting of 4 125 ms periods with a sampling rate of  44100 regardless of `format`.

```
access: RW_INTERLEAVED
format: S32_LE
subformat: STD
channels: 2
rate: 44100 (44100/1)
period_size: 5513
buffer_size: 22052
```
The above `period_size` and `buffer_size` values are in Frames.

Currently it looks like this with `S32` for example:

```
access: RW_INTERLEAVED
format: S32_LE
subformat: STD
channels: 2
rate: 44100 (44100/1)
period_size: 11026
buffer_size: 44104
```

**Reply by kingosticks** [reply]
Yes, you are right. The ALSA API used here works in frames and this code is now wrong.

The original code, before all this high resolution stuff came along, included comments which showed the period size calculated in bytes but then correctly set it in _frames_.

The old code wasn't perfect and the comment did have a slight error as well as setting the latency 4x too high... (it's the buffer size that matters there, rather than the period size, right?) so maybe that confused things when it all got changed.

**Reply by roderickvd** [reply]
> The old code wasn't perfect and the comment did have a slight error as well as setting the latency 4x too high... (it's the buffer size that matters there, rather than the period size, right?) so maybe that confused things when it all got changed.

Yes the error is mine, and that is probably what got me on the wrong foot.

@JasonLG1979 I just read through your code, will you open a PR? Checking the period size is a nice improvement too. As well as no longer requiring the `Frames` struct.

Only question I have is to change `PERIOD_TIME` into a `Duration::from_millis()` and later `PERIOD_TIME.as_micros() as u32`. As we discussed in the other PR, we're trying to make all the time constants a bit more Rust idiomatic. Thanks!

**Reply by JasonLG1979** [reply]
I can do that. I used `u32` because it was easier to wrap my non-corroded mind around (see what I did there,lol?) I will open a PR when I get it a presentable state.

**Reply by JasonLG1979** [reply]
@kingosticks:

> it's the buffer size that matters there, rather than the period size, right?

The buffer size is the total audio latency in Frames. Latency for playback is not really a big deal so long as it's under 1 sec (we shoot for about 0.5 sec). Network latency is more of a wildcard and maybe larger then the audio latency. A really high playback latency would just mean you might notice a delay when manually changing tracks or pausing and un-pausing.

Period size is the size of the periods, basically chunks of the buffer. The buffer is basically a segmented ring buffer. Period size is buffer size / periods. Not enough periods and you risk buffer underruns, to many periods and you hammer the CPU. Each period is an interrupt. 4 periods is the default for ALSA and seems to be what everyone uses as it seems like a happy medium between not getting underruns and not hammering the CPU.

**Reply by JasonLG1979** [reply]
@roderickvd:

> Checking the period size is a nice improvement too

Especially since `period_size`, `buffer_size`,  `buffer_time`, `period_time`, and `periods` can all be manually forced in `/etc/asound.conf`.

**Reply by JasonLG1979** [reply]
> Only question I have is to change PERIOD_TIME into a Duration::from_millis() and later PERIOD_TIME.as_micros() as u32. As we discussed in the other PR, we're trying to make all the time constants a bit more Rust idiomatic. Thanks!

Done.

**Comment by roderickvd** [comment]
Quick reply here while I’m gardening 😄 

I think you might be onto something here. One thing that crosses my mind is that in Rust Alsa, the `Frames` type is defined as a `c_long` which is `i32` or `i64` depending on the platform. How do you see this factoring in?

In the end the buffer should be big enough to hold the number of _bytes_ that we’re writing.

**Reply by JasonLG1979** [reply]
I think that `usize` makes sense for format size maybe? But not Frames. I think maybe that's one reason for the whole Frames abstraction? Any place in alsa.rs that takes plain number takes a `u32`? I don't know yet? I'll have to look at how big `self.buffer` actually is?

**Reply by roderickvd** [reply]
Yeah I thought `Frames` was pretty hawkish too just to mimic whatever `isize` is, when in fact you're not dealing with `Frames` at all. On my RPi I first changed it to some integer too, then saw that the CI failed on 64-bit hosts. I think your code fixes this because it no longer uses `hwp.set_period_size_near`. That function took `Frames` but the ones you're using now are using `u32`, much saner.

**Reply by JasonLG1979** [reply]
I just checked and the buffer size is `44096` bytes for S16 (if I have the unit correct?) in my code. To me that would be about right for a mono file for a full buffer (500ms) but I think that we might need to * `NUM_CHANNELS`?

Something like this maybe?

Edit: `periods` is already `u32`.

```rust
(period_frames as u32 * periods * NUM_CHANNELS as u32) as usize * format.size()
```

**Reply by JasonLG1979** [reply]
> the Frames type is defined as a c_long which is i32 or i64 depending on the platform. How do you see this factoring in?

My type conversion foo is not the greatest but I can't imagine frames being negative or being larger than `u32` I think it's pretty safe to convert `period_frames` `u32` as periods is already a `u32`?

**Reply by roderickvd** [reply]
> > the Frames type is defined as a c_long which is i32 or i64 depending on the platform. How do you see this factoring in?
> 
> My type conversion foo is not the greatest but I can't imagine frames being negative or being larger than `u32` I think it's pretty safe to convert `period_frames` `u32` as periods is already a `u32`?

Casting behavior well-defined in Rust: https://doc.rust-lang.org/nomicon/casts.html
In this situation I agree that it should be safe.

**Comment by JasonLG1979** [comment]
Dealing strictly in periods and time also opens up the possibility for a ```--buffer-size``` option in secs, `0.5` being the default? Although I can see it being used incorrectly. As people have it in their head that a really small buffer means really low latency when really low latency really doesn't matter that much for playback and can actually lead to more issues than it would solve if you go below `0.5`.

**Reply by roderickvd** [reply]
I think we should keep this KISS and let people override it in their backend, if so supported.

**Reply by JasonLG1979** [reply]
Thinking about it I tend to agree.

**Reply by kingosticks** [reply]
Agreed. People are obsessed with trying to minimise buffers like these and normally they have no idea what they are doing. Plus, less user options is better.

**Comment by roderickvd** [comment]
Closing this in favor of #789.
